src/transaction.ts

```diff
@@ -44,35 +63,47 @@ export function assembleTransaction(
     minAccountSequenceLedgerGap: raw.minAccountSequenceLedgerGap,
     extraSigners: raw.extraSigners,
   });
-  for (let i = 0; i < raw.operations.length; i++) {
-    const rawOp = raw.operations[i];
-    if ("function" in rawOp) {
-      const sim = simulated[i];
-      if (!sim) {
-        throw new Error("missing simulated operation");
-      }
-      let footprint = sim.footprint ?? rawOp.footprint;
-      if (!(footprint instanceof xdr.LedgerFootprint)) {
-        footprint = xdr.LedgerFootprint.fromXDR(footprint.toString(), "base64");
-      }
-      const auth = (sim.auth ?? rawOp.auth).map((a) =>
-        a instanceof xdr.ContractAuth
-          ? a
-          : xdr.ContractAuth.fromXDR(a.toString(), "base64"),
-      );
-      // TODO: Figure out a cleaner way to clone these operations
-      txn.addOperation(
-        Operation.invokeHostFunction({
-          function: rawOp.function,
-          parameters: rawOp.parameters,
-          footprint,
-          auth,
-        }),
-      );
-    } else {
-      // TODO: Handle this.
-      throw new Error("Unsupported operation type");
-    }
+
+  // apply the pre-built Auth from simulation onto each Tx/Op/HostFunction invocation
+  for (
+    let hostFnIndex = 0;
+    hostFnIndex < simulation.results.length;
+    hostFnIndex++
+  ) {
+    const rawHostFunction: xdr.HostFunction =
+      rawInvokeHostFunctionOp.functions[hostFnIndex];
+    const simHostFunctionResult: SorobanRpc.SimulateHostFunctionResult =
+      simulation.results[hostFnIndex];
+    rawHostFunction.auth(buildContractAuth(simHostFunctionResult.auth));
+    authDecoratedHostFunctions.push(rawHostFunction);
+  }
+
+  txnBuilder.addOperation(
+    Operation.invokeHostFunctions({
+      functions: authDecoratedHostFunctions,
+    }),
+  );
+
+  // apply the pre-built Soroban Tx Ext from simulation onto the Tx
+  txnBuilder.setExt(buildExt(simulation.transactionData));
+
+  return txnBuilder.build();
+}
+
+function buildExt(sorobanTxDataStr: string) {
+  const sorobanTxData: xdr.SorobanTransactionData = xdr.SorobanTransactionData.fromXDR(
+    sorobanTxDataStr,
+    "base64",
+  );
+  const txExt: xdr.TransactionExt = new xdr.TransactionExt(1);
```

Comment:
@Shaptic , do you have any insight on how to instantiate this TransactionExt union, i get a ts error on this constructor:

```
src/transaction.ts:98:60 - error TS2554: Expected 0 arguments, but got 1.

98   const txExt: xdr.TransactionExt = new xdr.TransactionExt(1);
```

it only happens from js-soroban-client, the same usage is working from [js-stellar-base in tx builder](https://github.com/stellar/js-stellar-base/pull/604/files#diff-38fee31808bd7d3eecb095e6b117fb4ea6c79843133da57fa7772f20de1fce6dR539), the one difference I can see is that the xdr is locally gen'd there rather than exported and then imported here. the exported version of it seems to be losing aspects of class construction.

Reply:
Oh jeez, I think the generator is busted on this? Can you do `xdr.TransactionExt[1]()`?

There's this in the Typescript:

```typescript
  class TransactionExt {
    switch(): number;

    sorobanData(value?: SorobanTransactionData): SorobanTransactionData;

    static 0(): TransactionExt;

    static 1(value: SorobanTransactionData): TransactionExt;

```

which is crazy: I don't think you can have only digits as a method name???

Reply:
yes, can't instantiate the class by anything other than no-args or single number, one indicator here related to ts is that getting the compile error on constructor method args, I can comment that away with @ts-ignore, and then it constructs, but the object is not good, if i call the `sorobanData(xdr.SorobanTransactionData)` method on it, it takes it but it doesn't update the union's arm value, subsequent calls to `.value()` return undefined, which is where the problem finally stops as can't marshal the tx to xdr now due to invalid object, getting `Error: XDR Write Error: undefined is not a SorobanTransactionData`.

I added transactionbuilder tests in the stellar-base pr to test this same behavior on the gen xdr but from js test:
https://github.com/stellar/js-stellar-base/pull/604/files#diff-2f0f2abfad28c352b947215436ec683c35dc35dbe4854a3a79f2b4be522779e5R70

and am seeing the same result, the constructed object doesn't work, and those new tests are failing with same error when they try to marshal back to xdr.

Reply:
@Shaptic , I think have narrowed the problem to js-xdr generated js for xdr unions with switches on number because the generated js for xdr unions with switch on types instead has correct js static factory methods based on the switch, and the objects generated by those work, I tested with:
```
let test = StellarBase.xdr.FeeBumpTransactionInnerTx.envelopeTypeTx(transaction.toEnvelope());
console.log("test value" + test.value());
```
the value returned matches the one passed on the static factory method, whereas for the gen'd code on number based switched unions, there is only that un-usable static factory method named with number, and using the default empty constructor does not work.

poking at js-sdk now ..

Reply:
Thanks for the bug report! We'll have to use the parsing workaround for now, since that generator fix might take some time (legacy codebase, nobody _really_ knows how it works :facepalm:)